### ISSUE TITLE
Create index concurrently

### ISSUE THREAD

**Issue by dnsbty** [tpetry/laravel-postgresql-enhanced#82]
For zero downtime migrations in Postgres, indexes should be created concurrently (outside of a transaction). It would be great if this extension could support that so that I don't have to drop down to raw SQL to make it happen.

Example migration query:
```sql
CREATE INDEX CONCURRENTLY idx_table_column ON public.table USING btree (column);
```

References:
https://www.postgresql.org/docs/current/sql-createindex.html#SQL-CREATEINDEX-CONCURRENTLY
https://medium.com/paypal-tech/postgresql-at-scale-database-schema-changes-without-downtime-20d3749ed680#b2f4

**Comment by tpetry** [comment]
Thats a good one! Sadly concurrent indexes can't be created within transactions. And every Laravel migration is automatically wrapped in transactions and I wouldn't want to suggest turning this off.

But I've thought about this feature also quite some time.

My idea was to create the index with another database connection. But if the current migration fails, the index will still be created. I had the idea of then probably killing the index creation. But I had not time yet to test whether dropping the index which has still not finished creating will also kill the current work its doing in the background.

Another problem is that the migration is finished but the index creation may take more minutes. So you probably want you migration process to also wait:

```php
$table->uniqueIndex('email')
  ->concurrently()
  ->wait();
```

So with the `->wait()` modifier the index would still be created concurrently in the background. But the migration would poll the database for the current state and only finish when the task is done. So steps happening after the migration would also wait for the index to exist.

**Comment by dnsbty** [comment]
The Migration class exposes a `$withinTransaction` variable that can be set to false to run a single migration without a transaction, and this is what I've done to make this work in migrations:

```php
<?php

use Illuminate\Database\Migrations\Migration;
use Illuminate\Support\Facades\DB;

return new class extends Migration {
    public $withinTransaction = false;

    public function up(): void
    {
        DB::statement(
            'CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_users_phone_number ON public.users USING btree (phone_number);'
        );
    }

    public function down(): void
    {
        DB::statement(
            'DROP INDEX CONCURRENTLY IF EXISTS idx_users_phone_number;'
        );
    }
};
```

Typically the `->wait()` modifier isn't something I would ever need. I would maybe expect to wait for an index to be created if it's on a brand new table and I know the index will be needed before accompanying code changes are deployed, but in those cases there won't be an issue with locking the table to create the index without the concurrently modifier. With the concurrently modifier, I'm typically adding an index to an existing table for performance reasons, and so my application is going to be running and using that data either way, so I have no need to block on the index being built.

Outside of Laravel, most frameworks tend to migrate indexes in this way. That same migration could be done in other frameworks like this:

Rails (notice `disable_ddl_transaction!`):
```rb
class AddIndexToUsersPhoneNumber < ActiveRecord::Migration
  disable_ddl_transaction!

  def change
    add_index :users, :phone_number, algorithm: :concurrently
  end
end
```

Django (notice `atomic = False`):
```py
import django.contrib.postgres.indexes
from django.db import migrations, models
from django.contrib.postgres.operations import AddIndexConcurrently


class Migration(migrations.Migration):

    atomic = False

    dependencies = [
        ("app_name", "001_initial"),
    ]

    operations = [
        AddIndexConcurrently(
            model_name="user",
            index=models.Index(
                fields=["name"], name="name_idx"
            ),
        ),
    ]
```

Phoenix (notice `@disable_ddl_transaction true`):
```ex
defmodule MyRepo.Migrations.CreateIndexOnUserPhoneNumber do
  use Ecto.Migration
  @disable_ddl_transaction true

  def change do
    create index("users", [:phone_number], concurrently: true)
  end
end
```

So I would recommend that we be able to take my above migration and do it like this:

```php
<?php

use Illuminate\Database\Migrations\Migration;
use Illuminate\Support\Facades\DB;

return new class extends Migration {
    public $withinTransaction = false;

    public function up(): void
    {
        Schema::table('users', function (Blueprint $table) {
            $table->index('phone_number')->concurrently();
        });
    }

    public function down(): void
    {
        Schema::table('leads', function (Blueprint $table) {
            $table->dropIndex(['phone_number'])->concurrently();
        });
    }
};
```

**Comment by tpetry** [comment]
I‘ll think about it. But I believe it also needs an `ifNotExists` modifier in case a migration (within a transaction) fails and is executed again. So the already existing will not be a problem.

**Comment by tpetry** [comment]
The feature is now available with release 0.39.0.

**Comment by dnsbty** [comment]
Thanks Tobias! This looks great!
